processor.go

```diff
@@ -442,6 +442,123 @@ func callInterfaceFunc(handler, input interface{}, streaming bool) (intf interfa
 	return
 }
 
+// ProcessClientStreamRequest allows clients to push multiple messages
+// asynchronously to the same service with the same connection. Unlike in
+// ProcessClientRequest, we take a channel of inputs that can be filled and will
+// subsequently call the service with any new messages received in the channel.
+func (p *ServiceProcessor) ProcessClientStreamRequest(req *http.Request, path string,
+	clientInputs chan []byte) ([]byte, *StreamingTunnel, error) {
+
+	outChan := make(chan []byte, 100)
+	buf := <-clientInputs
+	mh, ok := p.handlers[path]
+
+	reply, stopServiceChan, err := func() (interface{}, chan bool, error) {
+		if !ok {
+			err := xerrors.New("the requested message hasn't been " +
+				"registered: " + path)
+			log.Error(err)
+			return nil, nil, err
+		}
+
+		msg := reflect.New(mh.msgType).Interface()
+		err := protobuf.DecodeWithConstructors(buf, msg,
+			network.DefaultConstructors(p.Context.server.Suite()))
+		if err != nil {
+			return nil, nil, xerrors.Errorf("decoding: %v", err)
+		}
+
+		return callInterfaceFunc(mh.handler, msg, mh.streaming)
+	}()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// This goroutine is responsible for listening on the service channel,
+	// decoding the messages and then forwarding them to the streaming tunnel,
+	// which should then forward the message to the client.
+	go func() {
+		inChan := reflect.ValueOf(reply)
+		cases := []reflect.SelectCase{
+			reflect.SelectCase{Dir: reflect.SelectRecv, Chan: inChan},
+		}
+
+		for {
+			chosen, v, ok := reflect.Select(cases)
+			if !ok {
+				log.Lvlf4("publisher is closed for %s, closing "+
+					"outgoing channel", path)
+				close(outChan)
+				return
+			}
+			if chosen == 0 {
+				// Send information down to the client.
+				buf, err = protobuf.Encode(v.Interface())
+				if err != nil {
+					log.Error(err)
+					close(outChan)
+					return
+				}
+				outChan <- buf
+			} else {
+				panic("no such channel index")
+			}
+			// We don't add a way to explicitly stop the go-routine, otherwise
+			// the service will block. The service should close the channel when
+			// it has nothing else to say because it is the producer. Then this
+			// go-routine will be stopped as well.
+		}
+	}()
+
+	// This goroutine listens on any new messages from the client and executes
+	// the request. Executing the request should fill the service's channel, as
+	// the service will use the same chanel for further requests.
+	go func() {
+		for {
+			select {
+			case buf := <-clientInputs:
```

Comment:
If the channel is closed by the caller, it might panic in the decode function.

Reply:
surprisingly not: if the buffer is empty, the `DecodeWithConstructors` doesn't return an error, it just has no effect.

Reply:
the annoying thing is that it loops until the stopServiceChan is closed.

Reply:
so, should we say that as soon as there is an empty buffer we exit ?

Reply:
Or you can slightly change the API to not return the streaming tunnel but only the out channel and the caller has to close the client inputs channel when he's done.

Reply:
ok, I did that and it make total sense. We should be careful though, it's starting to look like real streaming...;) 

I don't know if we can keep it like that (probably not), but I then updated the `ProcessClientRequest` to reject streaming request, saying that it should instead use `ProcessClientStreamRequest`. Note that this does not affect the behavior of websocket.go